pkg/reconciler/autoscaling/kpa/kpa.go

```diff
@@ -239,24 +239,36 @@ func computeActiveCondition(ctx context.Context, pa *pav1alpha1.PodAutoscaler, p
 
 	switch {
 	case pc.want == 0:
-		if pa.Status.IsActivating() {
+		if pa.Status.IsActivating() && minReady > 0 {
 			// We only ever scale to zero while activating if we fail to activate within the progress deadline.
 			pa.Status.MarkInactive("TimedOut", "The target could not be activated.")
 		} else {
 			pa.Status.MarkInactive("NoTraffic", "The target is not receiving traffic.")
 		}
+		// This happens in the initialScale 0 case. After SKS has been setup, we override want -1
+		// with 0 when PA's desired scale is not set yet.
+		if pc.ready >= minReady {
+			pa.Status.MarkScaleTargetInitialized()
+		}
 
 	case pc.ready < minReady:
 		if pc.want > 0 || !pa.Status.IsInactive() {
 			pa.Status.MarkActivating(
 				"Queued", "Requests to the target are being buffered as resources are provisioned.")
+		} else {
+			// Need to set a condition because otherwise we will get NewObservedGenFailure because the reconciler
+			// does not set any condition during reconciliation of a new generation
+			pa.Status.MarkInactive("NoTraffic", "The target is not receiving traffic.")
 		}
 
 	case pc.ready >= minReady:
 		if pc.want > 0 || !pa.Status.IsInactive() {
-			pa.Status.MarkScaleTargetInitialized()
-			// SKS should already be active.
-			pa.Status.MarkActive()
+			// If initial scale is 0, we don't want to mark scale target initialized when
+			// want is still -1, because we will override -1 with 0 after SKS has been setup.
+			if minReady > 0 || (minReady == 0 && pc.want != -1) {
```

Comment:
I think you want to mention that pc.want is going to be -1 after first iteration when initialScale=0. Otherwise it's not clear how we end up in this situation.